README.md
# Building the chromeOS kernel
Use the build.sh file

## Manually(not recommended)
1. Clone the repository with git by running:
```bash
git clone --branch chromeos-5.10 --depth=1 https://chromium.googlesource.com/chromiumos/third_party/kernel.git
cd kernel
```

2. Download the kernel `.config` file, and update it, by running:
```bash
wget https://raw.githubusercontent.com/cb-linux/breath/main/kernel.conf -O .config
make olddefconfig
```

3. Compile the kernel by running:
```bash
make -j$(nproc)
```

The `bzImage` should be located in `arch/x86/boot/bzImage`.

Running `make -j$(nproc) modules_install INSTALL_MOD_PATH=[DIRECTORY]`, and then compressing the `[DIRECTORY]` should give you the compressed modules.


changelog.txt
Try newer stable version, release-R106-15054.B-chromeos-5.10

build.sh
#!/bin/bash

# install requirements
sudo apt update -y
sudo apt install -y netpbm imagemagick git build-essential ncurses-dev xz-utils libssl-dev bc flex libelf-dev bison cgpt vboot-kernel-utils

# Exit on errors
set -e

# If no kernel version in args
if [[ $# -eq 0 ]]; then
  printf "What kernel version would you like? (choose a number 1-3)"
  printf "This can be any ChromeOS kernel branch, but the recommended versions are:"
  printf "1: chromeos-5.10:\n  (Kernel version that is up-to-date and has good audio support with SOF)\n"
  printf "2: alt-chromeos-5.10:\n  (Kernel version that is up-to-date but is pinned to a commit that supports KBL/SKL devices which do not support SOF)\n"
  printf "3: chromeos-5.4:\n  (Often causes more issues despite being a commonly-used kernel in ChromeOS)\n"
  printf "Older kernels that may provide better support for specific boards:"
  printf "4: chromeos-4.19:\n  (For testing purposes)\n"
  printf "5: chromeos-4.14:\n  (For testing purposes)\n"
  printf "6: chromeos-4.4:\n  (For testing purposes; too old for Mesa3D and some other Linux userspace software)\n"
  printf "Newer kernels that are not widely used within ChromeOS devices:"
  printf "7: chromeos-5.15:\n  (Similar version to those used in Linux distributions, not used in any ChromeOS devices currently)\n"
  read -r KERNEL_VERSION
else
  export KERNEL_VERSION=$1
fi

export KERNEL_NAME=$KERNEL_VERSION

# Select correct release from Google
case $KERNEL_VERSION in
"1" | "chromeos-5.10") KERNEL_VERSION="release-R105-14989.B-chromeos-5.10" ;;
"2" | "alt-chromeos-5.10") KERNEL_VERSION="release-R86-13421.B-chromeos-5.4" ;;
"3" | "chromeos-5.4") KERNEL_VERSION="release-R101-14588.B-chromeos-5.4" ;;
"4" | "chromeos-4.19") KERNEL_VERSION="release-R101-14588.B-chromeos-4.19" ;;
"5" | "chromeos-4.14") KERNEL_VERSION="release-R101-14588.B-chromeos-4.14" ;;
"6" | "chromeos-4.4") KERNEL_VERSION="release-R101-14588.B-chromeos-4.4" ;;
"7" | "chromeos-5.15") KERNEL_VERSION="release-R106-15054.B-chromeos-5.15" ;;
*)
  printerr "Please supply a valid kernel version"
  exit
  ;;
esac

echo "Cloning kernel $KERNEL_VERSION"
if [[ ! -d $KERNEL_VERSION ]]; then
  if [[ $KERNEL_VERSION == "alt-chromeos-5.10" ]]; then
    git clone --branch chromeos-5.10 --single-branch https://chromium.googlesource.com/chromiumos/third_party/kernel.git $KERNEL_VERSION
    cd $KERNEL_VERSION
    git checkout "$(git rev-list -n 1 --first-parent --before="2021-08-1 23:59" chromeos-5.10)"
    cd ..
  else
    git clone --branch $KERNEL_VERSION --single-branch --depth 1 https://chromium.googlesource.com/chromiumos/third_party/kernel.git $KERNEL_VERSION
  fi
fi

(
  # TODO: fix logo
  echo "Setting up the bootlogo"
  cd logo
  mogrify -format ppm "logo.png"
  ppmquant 224 logo.ppm >logo_224.ppm
  pnmnoraw logo_224.ppm >logo_final.ppm
)

cd $KERNEL_VERSION
# CUSTOM EUPNEA PATCHES:
echo "Applying custom patches to the kernel"

cp ../logo/logo_final.ppm drivers/video/logo/logo_linux_clut224.ppm

# A somewhat commonly-used device, lesser priority than JSL i915
git apply ../patches/bloog-audio.patch || {
  git apply ../patches/bloog-audio.patch -R --check && echo "Bloog Audio Patch already applied"
}

# Super important patch, adds support for Jasperlake i915 (many DEDEDE devices)
# This is why we have a check to make sure again
git apply ../patches/jsl-i915.patch || {
  git apply ../patches/jsl-i915.patch -R --check && echo "Jasperlake iGPU Patch already applied"
}
grep -C3 "BIT(RCS0) | BIT(BCS0) | BIT(VCS0) | BIT(VECS0)" drivers/gpu/drm/i915/i915_pci.c | grep "jsl_info" -A5 | grep ".require_force_probe = 1" && {
  printerr "JSL Patch failed, exiting!"
  exit
}

# Important Jack Detection patch that fixes headphone jacks
git apply ../patches/jack-detection.patch || {
  git apply ../patches/jack-detection.patch -R --check && echo "Jack Detection Patch already applied"
}

# Utility functions not in the ChromeOS Kernel that are needed for the above patch to work
git apply ../patches/jack-detection-utils.patch || {
  git apply ../patches/jack-detection-utils.patch -R --check && echo "Jack Detection Utils Patch already applied"
}
echo "$(ls ../patches) applied"

# Prevents a dirty kernel
echo "mod" >>.gitignore
touch .scmversion

echo "Setting correct file names"
if [[ $KERNEL_NAME == "alt-chromeos-5.10" ]]; then
  MODULES="modules.alt.tar.xz"
  VMLINUZ="bzImage.alt"
  SYSTEM_MAP="System.map-eupnea-alt"
  CONFIG="config-eupnea-alt"
  [[ -f .config ]] || cp ../kernel.alt.conf .config || exit
elif [[ $KERNEL_NAME == "chromeos-5.15" ]]; then
  MODULES="modules.exp.tar.xz"
  VMLINUZ="bzImage.exp"
  SYSTEM_MAP="System.map-eupnea-exp"
  CONFIG="config-eupnea-exp"
  #[[ -f .config ]] || cp ../kernel.exp.conf .config || exit
  # using defautl kernel.conf for now
  [[ -f .config ]] || cp ../kernel.conf .config || exit
else
  MODULES="modules.tar.xz"
  VMLINUZ="bzImage"
  SYSTEM_MAP="System.map-eupnea"
  CONFIG="config-eupnea"
  [[ -f .config ]] || cp ../kernel.conf .config || exit
fi

echo "Making olddeconfig"
make olddefconfig

# If the terminal is interactive and not running in docker
if [[ -t 0 ]] && [[ ! -f /.dockerenv ]]; then
  read -p "Would you like to make edits to the kernel config? (y/n) " -n 1 -r
  echo
  if [[ $REPLY =~ ^[Yy]$ ]]; then
    make menuconfig
  fi

  read -p "Would you like to write the new config to github? (y/n) " -n 1 -r
  echo
  if [[ $REPLY =~ ^[Yy]$ ]]; then
    if [[ $KERNEL_VERSION == "alt-chromeos-5.10" ]]; then
      cp .config ../kernel.alt.conf
    else
      cp .config ../kernel.conf
    fi
  fi

  echo "Building kernel"
  read -p "Would you like a full rebuild? (y/n) " -n 1 -r
  echo
  if [[ $REPLY =~ ^[Yy]$ ]]; then
    make clean
    make -j"$(nproc)" || exit
  else
    make -j"$(nproc)" || exit
  fi
else
  make -j"$(nproc)"
fi

cp arch/x86/boot/bzImage ../$VMLINUZ
echo "bzImage and modules built"

rm -rf mod || true
mkdir mod
make -j"$(nproc)" modules_install INSTALL_MOD_PATH=mod
# Creates an archive containing /lib/modules/...

cd mod
# Speedy multicore compression
# Some version of tar don't support arguments after the command in the -I option,
# so we're putting the arguments and the command in a script
echo "xz -9 -T0" >fastxz
chmod +x fastxz
tar -cvI './fastxz' -f ../modules.tar.xz lib/
echo "modules.tar.xz created!"

# Copy the vmlinuz, modules.tar, system.map, and kernel config to the root directory
cd ..

# Print current path and files
echo "$MODULES"
echo "$SYSTEM_MAP"
echo "$CONFIG"

echo "$PWD"
echo $(sudo ls -a)
echo $(sudo ls -a ../)

cp modules.tar.xz ../$MODULES
cp System.map ../$SYSTEM_MAP
cp .config ../$CONFIG

echo "Command to extract modules to USB:"
echo "sudo rm -rf /mnt/lib/modules/* && sudo cp -Rv kernel/mod/lib/modules/* /mnt/lib/modules && sync"


kernel_build.py
#!/usr/bin/env python3
# CHROMEOS COMPILE INSTRUCTIONS: https://www.chromium.org/chromium-os/how-tos-and-troubleshooting/kernel-configuration/

import os
from functions import *
import argparse
import time
import json


# parse arguments from the cli.
def process_args():
    parser = argparse.ArgumentParser()
    parser.add_argument(dest="version", type=str, help="Kernel version to build(flag intended for docker containers)")
    parser.add_argument("-i", "--ignore-os", action="store_true", dest="ignore_os", default=False,
                        help="Allow building on non Ubuntu/debian based systems")
    return parser.parse_args()


def prepare_host() -> None:
    print("\033[96m" + "Preparing host system" + "\033[0m", flush=True)
    bash("apt update -y")
    bash("apt install -y netpbm imagemagick git build-essential ncurses-dev xz-utils libssl-dev bc flex libelf-dev "
         "bison binutils")
    # TODO: add cleaning after previous builds


def clone_kernel(kernel_head: str) -> None:
    print("\033[96m" + "Cloning kernel: " + kernel_head + "\033[0m", flush=True)
    if args.version == "alt-chromeos-5.10":
        bash("git clone --branch chromeos-5.10 --single-branch https://chromium.googlesource.com/chromiumos/third_party"
             "/kernel.git chromeos-kernel")
        # revert to older commit
        os.chdir("./chromeos-kernel")
        bash('git checkout "$(git rev-list -n 1 --first-parent --before="2021-08-1 23:59" chromeos-5.10)"')
    else:
        bash(f"git clone --branch {kernel_head} --single-branch --depth 1 https://chromium.googlesource.com/chromiumos"
             f"/third_party/kernel.git chromeos-kernel")
        os.chdir("./chromeos-kernel")


# Create boot icon/image
def create_boot_image():
    print("\033[96m" + "Creating boot image" + "\033[0m", flush=True)
    # TODO: Boot image


def apply_patches():
    print("\033[96m" + "Applying Eupnea patches" + "\033[0m", flush=True)

    print("\033[96m" + "Applying bloog audio patch" + "\033[0m", flush=True)
    try:
        bash_return(f"git apply ../patches/bloog-audio.patch")
    except subprocess.CalledProcessError:
        print("Bloog audio patch already applied", flush=True)

    print("\033[96m" + "Applying important jsl i915 patch" + "\033[0m", flush=True)
    try:
        bash_return(f"git apply ../patches/jsl-i915.patch")
    except subprocess.CalledProcessError:
        print("Checking if patch is already applied", flush=True)
        # check if patch is actually applied
        try:
            bash_return('grep -C3 "BIT(RCS0) | BIT(BCS0) | BIT(VCS0) | BIT(VECS0)" drivers/gpu/drm/i915/i915_pci.c | '
                        'grep "jsl_info" -A5 | grep -c ".require_force_probe = 1"')
        except subprocess.CalledProcessError:
            print("\033[91m" + "JSL i915 patch is not applied!! CRITICAL ERROR" + "\033[0m", flush=True)
        else:
            print("Bloog audio patch already applied", flush=True)

    print("\033[96m" + "Applying headphone jack patch" + "\033[0m", flush=True)
    try:
        bash_return(f"git apply ../patches/jack-detection.patch").strip()
    except subprocess.CalledProcessError:
        print("Headphone jack patch already applied", flush=True)

    print("\033[96m" + "Applying headphone jack utils patch" + "\033[0m", flush=True)
    try:
        bash_return(f"git apply ../patches/jack-detection-utils.patch").strip()
    except subprocess.CalledProcessError:
        print("Headphone jack utils patch already applied", flush=True)


def build_kernel() -> None:
    print("\033[96m" + "Preparing to build kernel" + "\033[0m", flush=True)
    # prevent dirty kernel build
    # add mod to .gitignore
    with open(".gitignore", "a") as file:
        file.write("mod")
    # create .scmversion
    with open(".scmversion", "w") as file:
        file.write("")

    # copy config file from GitHub
    try:
        rmfile(".config", )
    except FileNotFoundError:
        pass
    if args.version == "alt-chromeos-5.10":
        cpfile("../kernel-alt.conf", "./.config")
    else:
        cpfile("../kernel.conf", "./.config")

    # make config with default selections
    print("\033[96m" + "Making config with default options" + "\033[0m", flush=True)
    bash("make olddefconfig")
    # TODO: add config editing for users

    print("\033[96m" + "Building kernel" + "\033[0m", flush=True)
    kernel_start = time.time()
    try:
        bash(f"make -j{cores}")
    except subprocess.CalledProcessError:
        print("\033[91m" + f"Kernel build failed in: {time.time() - kernel_start}" + "\033[0m")
        exit(1)
    print("\033[96m" + f"Kernel build succeeded in: {time.time() - kernel_start}" + "\033[0m", flush=True)


def build_modules() -> None:
    print("\033[96m" + "Preparing for modules build" + "\033[0m", flush=True)
    rmdir("mod")  # just in case
    mkdir("mod")

    print("\033[96m" + "Building modules" + "\033[0m", flush=True)
    modules_start = time.time()
    try:
        bash(f"make -j{cores} modules_install INSTALL_MOD_PATH=mod")
    except subprocess.CalledProcessError:
        print("\033[91m" + f"Modules build failed in: {time.time() - modules_start}" + "\033[0m")
        exit(1)
    print("\033[96m" + f"Modules build succeeded in: {time.time() - modules_start}" + "\033[0m", flush=True)

    print("\033[96m" + "Compressing modules" + "\033[0m", flush=True)
    os.chdir("./mod")
    # TODO: convert to one liner
    # create extraction script
    with open("fastxz", "w") as file:
        file.write("xz -9 -T0")
    bash("chmod +x fastxz")  # make script executable
    modules_start = time.time()
    try:
        bash("tar -cvI './fastxz' -f ../modules.tar.xz lib/")
    except subprocess.CalledProcessError:
        print("\033[91m" + f"Modules archival failed in: {time.time() - modules_start}" + "\033[0m")
        exit(1)
    print("\033[96m" + f"Modules archival succeeded in: {time.time() - modules_start}" + "\033[0m", flush=True)
    os.chdir("..")  # go back to chromeos kernel root


if __name__ == "__main__":
    # Elevate script to root
    if not os.geteuid() == 0:
        sudo_args = ['sudo', sys.executable] + sys.argv + [os.environ]
        os.execlpe('sudo', *sudo_args)
    script_start = time.time()
    args = process_args()

    # get number of cores
    cores = bash_return("nproc")
    print(f"Available cores: {cores}", flush=True)

    # check if running on ubuntu
    if not path_exists("/usr/bin/apt") and args.ignore_os:  # check if running on ubuntu/debian
        print("This script is made for Ubuntu(docker). Use --ignore-os to run on other systems.\n"
              " Install these packages using your package manager:", flush=True)
        print("netpbm imagemagick git build-essential ncurses-dev xz-utils libssl-dev bc flex libelf-dev bison cgpt "
              "vboot-kernel-utils", flush=True)
        exit(1)
    if args.version == "":
        print("Which kernel version would you like to use? ", flush=True)
        print("Manual building is not supported yet. Use the old script for now.", flush=True)
        exit(1)
        # TODO: add version selection for users
    prepare_host()
    # get kernel_head
    try:
        with open("kernel_versions.json", "r") as json_file:
            read_kernel_head = json.load(json_file)[args.version]
    except KeyError:
        print("\033[91m" + "Kernel version not available." + "\033[0m", flush=True)
        exit(1)
    clone_kernel(read_kernel_head)
    create_boot_image()
    apply_patches()
    build_kernel()
    build_modules()

    # determine file names
    match args.version:
        case "alt-chromeos-5.10":
            modules_name = "modules-alt.tar.xz"
            bzImage_name = "bzImage-alt"
            # system_map_name = "System-alt.map"
            config_name = "kernel-alt.config"
        case "chromeos-5.15":
            modules_name = "modules-exp.tar.xz"
            bzImage_name = "bzImage-exp"
            #  system_map_name = "System-exp.map"
            config_name = "kernel-exp.config"
        case "chromeos-5.10":
            modules_name = "modules.tar.xz"
            bzImage_name = "bzImage"
            #  system_map_name = "System.map"
            config_name = "kernel.config"
        case _:
            # shouldn't be built by GitHub actions
            modules_name = "modules-old.tar.xz"
            bzImage_name = "bzImage-old"
            # system_map_name = "System-old.map"
            config_name = "kernel-old.config"

    # copy files to actual root
    print("\033[96m" + "Copying files to actual root" + "\033[0m", flush=True)

    cpfile("modules.tar.xz", f"../{modules_name}")
    cpfile("arch/x86/boot/bzImage", f"../{bzImage_name}")
    # cp("System.map", f"../{system_map_name}")
    cpfile(".config", f"../{config_name}")

    print("\033[96m" + f"Build completed in: {time.time() - script_start}" + "\033[0m", flush=True)
